lib/audio/music_service.dart

```diff
@@ -0,0 +1,102 @@
+import 'dart:async';
+
+import 'package:audioplayers/audioplayers.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter_game_challenge/settings/settings.dart';
+
+class MusicService {
+  final AudioPlayer _musicPlayer;
+  final AudioPlayer Function() _createAudioPlayer;
+  final SettingsController _settingsController;
+  bool _isMusicEnabled;
+  bool _isSoundsEnabled;
+  AssetSource? _latestSource;
+
+  MusicService(
+    this._createAudioPlayer,
+    this._settingsController,
+  )   : _isMusicEnabled = _settingsController.musicOn.value,
+        _isSoundsEnabled = _settingsController.soundsOn.value,
+        _musicPlayer = _createAudioPlayer() {
+    _settingsController.musicOn.addListener(_updateIsMusicEnabled);
+    _settingsController.soundsOn.addListener(_updateIsSoundsEnabled);
+  }
+
+  ValueNotifier<AppLifecycleState>? _lifecycleNotifier;
+  StreamSubscription<void>? _songCompletedSubscription;
+
+  set lifecycleNotifier(ValueNotifier<AppLifecycleState> value) {
+    _lifecycleNotifier = value;
+
+    _lifecycleNotifier?.addListener(() async {
+      final appState = _lifecycleNotifier?.value;
+
+      appState != AppLifecycleState.resumed
+          ? await _musicPlayer.pause()
+          : await _musicPlayer.resume();
+    });
+  }
+
+  Future<void> stopMusic() {
+    return _musicPlayer.stop();
+  }
+
+  Future<void> playSound(AssetSource source) async {
+    if (!_isSoundsEnabled) {
+      return;
+    }
+
+    final tempPlayer = _createAudioPlayer();
+    await tempPlayer.play(source);
+
+    StreamSubscription<void>? tempSubscription;
+    tempSubscription = tempPlayer.onPlayerComplete.listen((event) {
+      tempSubscription?.cancel();
+      tempPlayer.dispose();
+    });
```

Comment:
From my point of view we need to have a dedicated variable for _tempSubscription in the MusicService and set it before play method, cause currently we are subscribing to the tempPlayer events after its have played a sound.

Reply:
Need to discuss

Reply:
Discussed in pm, there is no need for refactoring here.